### DISCUSSION TITLE
Conditional abbreviations

### DISCUSSION THREAD

**Discussion by rbhanot4739** [olets/zsh-abbr#158] [Q&A] [answered]
Thank you for the amazing plugin, and it works really well. Just wondering if its possible to add certain abbreviations only if the command exists. For example, this is what I have in my aliases file

```sh
if [[ $(command -v eza) ]]; then
	alias ls='eza -I "*pyc*" $eza_params'
	alias lsa='ls --all'
        ...
fi
```
However, when I import these aliases into abbreviations, this logic isn't respected automatically (which is acceptable and not a big deal). So is it possible to add such `if/else` conditions in the `$ABBR_USER_ABBREVIATIONS_FILE` and have an abbreviation defined based on that ?

**Comment by olets** [comment, marked as answer]
Interesting question!

Three solutions come to mind:

Put the condition in the abbreviated command itself (wrapping the `if` condition in `{}` prevents its exit code from affecting whether the `else` is reached).

```shell
% abbr ls='command -v eza && { eza -I "*pyc*" $eza_params } || ls'
```

Or use the standard command as a fallback (`2>/dev/null` prevents the "zsh: command not found" output from showing in your terminal)

```shell
% abbr ls='eza -I "*pyc*" $eza_params 2>/dev/null || ls'
```

Or conditionally add ~[session abbreviations](https://zsh-abbr.olets.dev/scopes.html#session)~ _edit: abbreviations_ when the shell starts:

```shell
# .zshrc

# load zsh-abbr and then

if …; then
  # edit: this is what I said originally
  # abbr -S …
  # but in retrospect (see henrebotha's response) this is probably preferable
  abbr …
fi
```

**Reply by henrebotha** [reply]
I set up abbreviations (not session abbreviations though) [on shell startup](https://github.com/henrebotha/dotfiles/blob/b0eca753b16635da4db81e114451e412a4795519/zsh/.config/zsh/.zshrc#L460), like option 3, which works great except for clobbering my p10k prompt a little bit when it fires because I can't be bothered to learn which p10k setting will prevent that. Anyway, because it's not session-specific, it only fires when there's something new to abbreviate, i.e. once per new machine setup and/or dotfiles-tinkering day.

So I'd personally go for something like option 3.

**Reply by olets** [reply]
Belatedly: thanks @henrebotha. Looking back at this, I don't see a reason to use session abbreviations in option 3.

**Reply by henrebotha** [reply]
I've since started using chezmoi, which gives me a better way to implement this that adds no shell startup time or anything.
